presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UncorrelatedInPredicateApplyRemover.java

```diff
@@ -0,0 +1,282 @@
+                builder.putAll(node.getAssignments());
+                builder.put(symbol, symbol.toQualifiedNameReference());
+                return new ProjectNode(node.getId(), node.getSource(), builder.build());
+            }
+            else {
+                return node;
+            }
+        }
+
+        @Override
+        public PlanNode visitApply(ApplyNode node, RewriteContext<InsertSemiJoinContext> context)
+        {
+            InPredicate inPredicate = context.get().peekInPredicate();
+            if (inPredicate.getValueList() instanceof QualifiedNameReference) {
+                Symbol value = asSymbol(inPredicate.getValue());
+                Symbol valueList = asSymbol(inPredicate.getValueList());
+                if (node.getInput().getOutputSymbols().contains(value) && node.getSubquery().getOutputSymbols().contains(valueList)) {
+                    Symbol semijoinResult = symbolAllocator.newSymbol("semijoin_result", BOOLEAN);
+                    context.get().setInPredicateSymbol(semijoinResult);
+                    return new SemiJoinNode(idAllocator.getNextId(),
+                            node.getInput(),
+                            node.getSubquery(),
+                            value,
+                            valueList,
+                            semijoinResult,
+                            Optional.empty(),
+                            Optional.empty()
+                    );
+                }
+                else {
+                    return new ApplyNode(node.getId(),
+                            context.rewrite(node.getInput(), context.get()),
+                            node.getSubquery(),
+                            node.getCorrelation());
+                }
+            }
+            else {
+                return super.visitApply(node, context);
+            }
+        }
+
+        private Symbol asSymbol(Expression expression)
+        {
+            checkState(expression instanceof QualifiedNameReference);
+            return Symbol.fromQualifiedName(((QualifiedNameReference) expression).getName());
+        }
+    }
+
+    private class InsertSemiJoinContext
+    {
+        private LinkedList<InPredicate> inPredicates = new LinkedList<>();
+        private Optional<Symbol> inPredicateSymbol = Optional.empty();
+
+        public InsertSemiJoinContext(InPredicate inPredicate)
+        {
+            pushInPredicate(inPredicate);
+        }
+
+        public void pushInPredicate(InPredicate inPredicate)
+        {
+            this.inPredicates.addLast(inPredicate);
+        }
+
+        public InPredicate peekInPredicate()
+        {
+            checkState(!inPredicates.isEmpty(), "InPredicate is not set");
+            return inPredicates.getLast();
+        }
+
+        public void setInPredicateSymbol(Symbol symbol)
+        {
+            this.inPredicateSymbol = Optional.of(symbol);
+        }
+
+        public Map<InPredicate, Expression> getInPredicateMapping()
```

Comment:
what you really care here is just top level IN predicate. Maybe it would be better to return just one entry here - not whole map.